File: src/utils/imageUtils.ts
const VALID_IMAGE_TYPES = new Set([
  'image/png',
  'image/jpeg',
  'image/jpg',
  'image/gif',
  'image/webp',
  'image/bmp',
])

const VALID_IMAGE_EXTENSIONS = new Set(['png', 'jpg', 'jpeg', 'gif', 'webp', 'bmp'])

const DEFAULT_MAX_WIDTH = 2048
const DEFAULT_MAX_HEIGHT = 2048
const DEFAULT_QUALITY = 0.8
const DEFAULT_MAX_SIZE_MB = 10

export interface ImageDimensions {
  width: number
  height: number
}

export interface ValidationResult {
  valid: boolean
  error?: string
}

export function isImageFile(file: File): boolean {
  if (VALID_IMAGE_TYPES.has(file.type)) return true
  const ext = file.name.split('.').pop()?.toLowerCase() ?? ''
  return VALID_IMAGE_EXTENSIONS.has(ext)
}

export function validateImage(
  file: File,
  maxSizeMB: number = DEFAULT_MAX_SIZE_MB,
): ValidationResult {
  if (!isImageFile(file)) {
    return {
      valid: false,
      error: `Unsupported image type: ${file.type || 'unknown'}. Supported: PNG, JPG, GIF, WebP, BMP`,
    }
  }
  if (file.size > maxSizeMB * 1024 * 1024) {
    return {
      valid: false,
      error: `Image is ${(file.size / (1024 * 1024)).toFixed(1)}MB, exceeding the ${maxSizeMB}MB limit`,
    }
  }
  return { valid: true }
}

export function getImageDimensions(file: File): Promise<ImageDimensions> {
  return new Promise((resolve, reject) => {
    const img = new Image()
    const url = URL.createObjectURL(file)
    img.onload = () => {
      resolve({ width: img.naturalWidth, height: img.naturalHeight })
      URL.revokeObjectURL(url)
    }
    img.onerror = () => {
      URL.revokeObjectURL(url)
      reject(new Error(`Failed to load image: ${file.name}`))
    }
    img.src = url
  })
}

export function compressImage(
  file: File,
  maxWidth: number = DEFAULT_MAX_WIDTH,
  maxHeight: number = DEFAULT_MAX_HEIGHT,
  quality: number = DEFAULT_QUALITY,
): Promise<Blob> {
  return new Promise((resolve, reject) => {
    const img = new Image()
    const url = URL.createObjectURL(file)
    img.onload = () => {
      URL.revokeObjectURL(url)
      let { naturalWidth: w, naturalHeight: h } = img

      if (w > maxWidth || h > maxHeight) {
        const ratio = Math.min(maxWidth / w, maxHeight / h)
        w = Math.round(w * ratio)
        h = Math.round(h * ratio)
      }

      const canvas = document.createElement('canvas')
      canvas.width = w
      canvas.height = h
      const ctx = canvas.getContext('2d')
      if (!ctx) {
        reject(new Error('Failed to get canvas 2d context'))
        return
      }
      ctx.drawImage(img, 0, 0, w, h)

      const outputType = file.type === 'image/png' ? 'image/png' : 'image/jpeg'
      const outputQuality = outputType === 'image/png' ? undefined : quality

      canvas.toBlob(
        (blob) => {
          if (blob) {
            resolve(blob)
          } else {
            reject(new Error(`Failed to compress image: ${file.name}`))
          }
        },
        outputType,
        outputQuality,
      )
    }
    img.onerror = () => {
      URL.revokeObjectURL(url)
      reject(new Error(`Failed to load image for compression: ${file.name}`))
    }
    img.src = url
  })
}

export function imageToBase64(file: File | Blob): Promise<string> {
  return new Promise((resolve, reject) => {
    const reader = new FileReader()
    reader.onload = () => {
      const result = reader.result
      if (typeof result === 'string') {
        resolve(result)
      } else {
        reject(new Error('FileReader did not return a string'))
      }
    }
    reader.onerror = () => {
      reject(
        new Error(`Failed to read file as base64: ${file instanceof File ? file.name : 'Blob'}`),
      )
    }
    reader.readAsDataURL(file)
  })
}

export async function processImageForSending(
  file: File,
  maxWidth: number = DEFAULT_MAX_WIDTH,
  maxHeight: number = DEFAULT_MAX_HEIGHT,
  quality: number = DEFAULT_QUALITY,
  maxSizeMB: number = DEFAULT_MAX_SIZE_MB,
): Promise<string> {
  const validation = validateImage(file, maxSizeMB)
  if (!validation.valid) {
    throw new Error(validation.error)
  }

  const compressed = await compressImage(file, maxWidth, maxHeight, quality)
  return imageToBase64(compressed)
}

export function extractBase64Data(dataUri: string): { mediaType: string; base64: string } {
  const match = dataUri.match(/^data:([^;]+);base64,(.+)$/)
  if (!match) {
    throw new Error('Invalid data URI format')
  }
  return { mediaType: match[1], base64: match[2] }
}

File: frontend/src/utils/image_utils.rs
//! Image utility functions — ported from src/utils/imageUtils.ts
//!
//! The TS version uses browser APIs (File, Image, Canvas, FileReader).
//! This Rust version provides equivalent logic for validation and base64
//! extraction, while image compression/resize would use the `image` crate
//! on the Tauri backend side.

use once_cell::sync::Lazy;
use std::collections::HashSet;

/// Default maximum image width for compression.
pub const DEFAULT_MAX_WIDTH: u32 = 2048;

/// Default maximum image height for compression.
pub const DEFAULT_MAX_HEIGHT: u32 = 2048;

/// Default JPEG quality for compression (0.0–1.0).
pub const DEFAULT_QUALITY: f32 = 0.8;

/// Default maximum file size in megabytes.
pub const DEFAULT_MAX_SIZE_MB: u32 = 10;

static VALID_IMAGE_TYPES: Lazy<HashSet<&'static str>> = Lazy::new(|| {
    [
        "image/png",
        "image/jpeg",
        "image/jpg",
        "image/gif",
        "image/webp",
        "image/bmp",
    ]
    .iter()
    .cloned()
    .collect()
});

static VALID_IMAGE_EXTENSIONS: Lazy<HashSet<&'static str>> = Lazy::new(|| {
    ["png", "jpeg", "jpg", "gif", "webp", "bmp"]
        .iter()
        .cloned()
        .collect()
});

/// Result of image validation.
#[derive(Debug, Clone)]
pub struct ValidationResult {
    pub valid: bool,
    pub error: Option<String>,
}

/// Check if a MIME type or file extension indicates an image file.
pub fn is_image_file(mime_type: Option<&str>, filename: Option<&str>) -> bool {
    if let Some(mt) = mime_type {
        if VALID_IMAGE_TYPES.contains(mt) {
            return true;
        }
    }
    if let Some(name) = filename {
        let ext = name.rsplit('.').next().unwrap_or("").to_lowercase();
        return VALID_IMAGE_EXTENSIONS.contains(ext.as_str());
    }
    false
}

/// Validate an image by MIME type, extension, and size.
pub fn validate_image(
    mime_type: Option<&str>,
    filename: Option<&str>,
    size_bytes: u64,
    max_size_mb: Option<u32>,
) -> ValidationResult {
    let max_mb = max_size_mb.unwrap_or(DEFAULT_MAX_SIZE_MB);
    if !is_image_file(mime_type, filename) {
        return ValidationResult {
            valid: false,
            error: Some("File is not a supported image type.".to_string()),
        };
    }
    if size_bytes > (max_mb as u64) * 1024 * 1024 {
        return ValidationResult {
            valid: false,
            error: Some(format!("Image exceeds maximum size of {}MB.", max_mb)),
        };
    }
    ValidationResult {
        valid: true,
        error: None,
    }
}

/// Parsed base64 data from a data URI.
#[derive(Debug, Clone)]
pub struct Base64Data {
    pub media_type: String,
    pub base64: String,
}

/// Extract the media type and base64 payload from a data URI
/// (`data:<mediatype>;base64,<data>`).
pub fn extract_base64_data(data_uri: &str) -> Result<Base64Data, String> {
    if !data_uri.starts_with("data:") {
        return Err("Invalid data URI: must start with 'data:'".to_string());
    }
    let rest = &data_uri[5..];
    let (header, data) = rest
        .split_once(',')
        .ok_or("Invalid data URI: missing comma separator")?;
    let media_type = header
        .strip_suffix(";base64")
        .ok_or("Invalid data URI: missing ';base64' suffix")?
        .to_string();
    Ok(Base64Data {
        media_type,
        base64: data.to_string(),
    })
}
